File: C++/stochastic.cpp
#include <cmath>
#include "stochastic.h"
#include "statistics.h"

/**
 * Calculate European option price using Bachelier model:
 * dSt = sigma * S0 * dWt 
 * St = S0*(1 + sigma*Wt)
 * 
 * @param Spot price of underlying asset at time 0.
 * @param Strike strike price of option
 * @param Vol variance of Brownian motion
 * @param T length of time
 * @param type type of European option. Including: van call/put (vanilla), con call/put (cash-or-nothing), aon call/put (asset-or-nothing)
 * @return value of the option at time 0 
 */
double Bachelier(double Spot, double Strike, double Vol, double T, std::string type)
{
    double val;
    double xs = (Strike - Spot) / (Spot * Vol * sqrt(T));
    StandardNormalDistribution Z;

    if (Spot == Strike)
    {
        return Vol * Spot * sqrt(0.5 * T / M_PI);
    }

    if (type == "van call")
    {
        val = (Spot - Strike) * Z.cdf(-xs) + Spot * Vol * sqrt(T) * Z.pdf(-xs);
    }
    else if (type == "van put")
    {
        val = (Strike - Spot) * Z.cdf(xs) + Spot * Vol * sqrt(T) * Z.pdf(xs);
    }
    else if (type == "con call")
    {
        val = Z.cdf(-xs);
    }
    else if (type == "con put")
    {
        val = Z.cdf(xs);
    }
    else if (type == "aon call")
    {
        val = Spot * Z.cdf(-xs) + Spot * Vol * sqrt(T) * Z.pdf(-xs);
    }
    else if (type == "aon put")
    {
        val = Spot * Z.cdf(xs) - Spot * Vol * sqrt(T) * Z.pdf(xs);
    }
    else
    {
        throw std::invalid_argument("Option type is invalid. Check documentation");
    }

    return val;
}

/**
 * Calculate European option price using Black-Scholes (1973) model:
 * dSt = r*dSt + sigma*St*dWt 
 * St = S0*exp{(r-sigma^2/2)t + sigma*Wt}
 * 
 * @param Spot price of underlying asset at time 0.
 * @param Strike strike price of option
 * @param r drift of St
 * @param Vol variance of Brownian motion
 * @param T length of time
 * @param type type of European option. Including: van call/put (vanilla), con call/put (cash-or-nothing), aon call/put (asset-or-nothing)
 * @return value of the option at time 0 
 */
double BlackScholes(double Spot, double Strike, double r, double Vol, double T, std::string type)
{
    double val;
    double d1 = (log(Spot / Strike) + (r + 0.5 * Vol * Vol) * T) / (Vol * sqrt(T));
    double d2 = (log(Spot / Strike) + (r - 0.5 * Vol * Vol) * T) / (Vol * sqrt(T));
    StandardNormalDistribution Z;

    if (Spot == Strike)
    {
        return Vol * Spot * sqrt(0.5 * T / M_PI);
    }

    if (type == "van call")
    {
        val = Spot * Z.cdf(d1) - Strike * exp(-r * T) * Z.cdf(d2);
    }
    else if (type == "van put")
    {
        val = -Spot * Z.cdf(-d1) + Strike * exp(-r * T) * Z.cdf(-d2);
    }
    else if (type == "con call")
    {
        val = exp(-r * T) * Z.cdf(d2);
    }
    else if (type == "con put")
    {
        val = exp(-r * T) * Z.cdf(-d2);
    }
    else if (type == "aon call")
    {
        val = Spot * Z.pdf(d1);
    }
    else if (type == "aon put")
    {
        val = Spot * Z.pdf(-d1);
    }
    else
    {
        throw std::invalid_argument("Option type is invalid. Check documentation");
    }

    return val;
}

File: Python/stochastic.py
#!/usr/bin/env python
import numpy as np
from scipy.stats import norm

def bachelier(So, K, sigma, T, option_type):
    '''
    Calculate European option price using Bachelier model:
        dSt = sigma * S0 * dWt 
        St = S0*(1 + sigma*Wt)
    
    Parameter
    ---------
    So: float
        price of underlying asset at time 0
    K: float
        strike price of option
    sigma: float
        variance of Brownian motion
    T: float
        length of time
    option_type: str
        type of European option. 
        Including: van call/put (vanilla), con call/put (cash-or-nothing), aon call/put (asset-or-nothing)
    
    Return
    ------
    val: value of the option at time 0
    '''
    
    
    xs = (K-So) / (So * sigma * np.sqrt(T))
    val = None
    
    if So == K:
        return sigma*So*np.sqrt(T/(2*np.pi))
    
    if option_type == 'van call':
        val = (So - K) * norm.cdf(-xs) + So*sigma*np.sqrt(T)*norm.pdf(-xs)
    elif option_type == 'van put':
        val = (K - So) * norm.cdf(xs) + So*sigma*np.sqrt(T)*norm.pdf(xs)
    elif option_type == 'con call':
        val = norm.cdf(-xs)
    elif option_type == 'con put':
        val = norm.cdf(xs)
    elif option_type == 'aon call':
        val = So*norm.cdf(-xs) + So*sigma*np.sqrt(T)*norm.pdf(-xs)
    elif option_type == 'aon put':
        val = So*norm.cdf(xs) - So*sigma*np.sqrt(T)*norm.pdf(xs)
    else:
        raise(ValueError("Option type is invalid. " +
                         "Should be either 'van call', 'van put', 'con call', 'con put', 'aon call', or 'aon put'"))
    
    return val
                
    
def black_scholes(So, K, r, sigma, T, option_type):
    '''
    Calculate European option price using Black-Scholes (1973) model:
        dSt = r*dSt + sigma*St*dWt 
        St = S0*exp{(r-sigma^2/2)t + sigma*Wt}
    
    Parameter
    ---------
    So: float
        price of underlying asset at time 0
    K: float
        strike price of option
    r: float
        drift of St
    sigma: float
        variance of Brownian motion
    T: float
        length of time
    option_type: str
        type of European option. 
        Including: van call/put (vanilla), con call/put (cash-or-nothing), aon call/put (asset-or-nothing)
    
    Return
    ------
    val: value of the option at time 0
    '''
    
    
    d1 = (np.log(So/K) + (r+sigma**2/2)*T) / (sigma*np.sqrt(T))
    d2 = (np.log(So/K) + (r-sigma**2/2)*T) / (sigma*np.sqrt(T))
    val = None
    
    if So == K:
        return sigma*So*np.sqrt(T/(2*np.pi))
    
    if option_type == 'van call':
        val = So*norm.cdf(d1) - K*np.e**(-r*T)*norm.cdf(d2)
    elif option_type == 'van put':
        val = -So*norm.cdf(-d1) + K*np.e**(-r*T)*norm.cdf(-d2)
    elif option_type == 'con call':
        val = np.e**(-r*T) * norm.cdf(d2)
    elif option_type == 'con put':
        val = np.e**(-r*T) * norm.cdf(-d2)
    elif option_type == 'aon call':
        val = So*norm.cdf(d1)
    elif option_type == 'aon put':
        val = So*norm.cdf(-d1)
    else:
        raise(ValueError("Option type is invalid. " +
                         "Should be either 'van call', 'van put', 'con call', 'con put', 'aon call', or 'aon put'"))
    
    return val
                